mindmeld/models/taggers/pytorch_crf.py

```diff
@@ -0,0 +1,407 @@
+
+from ...exceptions import MindMeldError
+from ...path import USER_CONFIG_DIR
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_PYTORCH_CRF_ER_CONFIG = {
+    "feat_type": "hash",  # ["hash", "dict"]
+    "feat_num": 50000,
+    "stratify": True,
+    "drop_input": 0.2,
+    "train_batch_size": 8,
+    "patience": 3,
+    "epochs": 100,
+    "train_dev_split": 0.15,
+    "optimizer_type": "sgd",  # ["sgd", "adam"]
+}
+
+
+class TaggerDataset(Dataset):
+    def __init__(self, inputs, seq_lens, labels=None):
+        self.inputs = inputs
+        self.labels = labels
+        self.seq_lens = seq_lens
+        self.max_seq_length = max(seq_lens)
+
+    def __len__(self):
+        return len(self.seq_lens)
+
+    def __getitem__(self, index):
+        mask_list = [1] * self.seq_lens[index] + [0] * (self.max_seq_length - self.seq_lens[index])
+
+        mask = torch.as_tensor(mask_list, dtype=torch.bool)
+        if self.labels:
+            return self.inputs[index], mask, self.labels[index]
+
+        return self.inputs[index], mask
+
+
+def custom_coo_cat(tensors):
+    assert len(tensors) > 0
+
+    rows = []
+    cols = []
+    values = []
+    sparse_sizes = [0, 0]
+
+    nnz = 0
+    for tensor in tensors:
+        tensor = tensor.coalesce()
+        row, col = tensor.indices()[0], tensor.indices()[1]
+        if row is not None:
+            rows.append(row + sparse_sizes[0])
+
+        cols.append(col + sparse_sizes[1])
+
+        value = tensor.values()
+        if value is not None:
+            values.append(value)
+
+        sparse_sizes[0] += tensor.shape[0]
+        sparse_sizes[1] += tensor.shape[1]
+        nnz += tensor._nnz()
+
+    row = None
+    if len(rows) == len(tensors):
+        row = torch.cat(rows, dim=0)
+
+    col = torch.cat(cols, dim=0)
+
+    value = None
+    if len(values) == len(tensors):
+        value = torch.cat(values, dim=0)
+
+    return torch.sparse_coo_tensor(indices=torch.stack([row, col]), values=value, size=sparse_sizes).coalesce()
+
+
+def init_weights(m):
+    if isinstance(m, nn.Linear):
+        torch.nn.init.xavier_normal_(m.weight)
+        m.bias.data.fill_(0.01)
+
+
+def custom_collate(sequence):
+    if len(sequence[0]) == 3:
+        sparse_mats, masks, labels = zip(*sequence)
+        return custom_coo_cat(sparse_mats), torch.stack(masks), torch.stack(labels)
+    elif len(sequence[0]) == 2:
+        sparse_mats, masks = zip(*sequence)
+        return custom_coo_cat(sparse_mats), torch.stack(masks)
+
+
+class Encoder:
+    def __init__(self, feature_extractor="hash", num_feats=50000):
+
+        self.feat_extractor = DictVectorizer(dtype=np.float32) if feature_extractor == "dict" else FeatureHasher(
+            n_features=num_feats, dtype=np.float32)
+        self.label_encoder = LabelEncoder()
+        self.feat_extract_type = feature_extractor
+        self.num_classes = None
+        self.classes = None
+        self.num_feats = num_feats
+        self.fit_done = False
```

Comment:
To keep things consistent across the repo, can we change this to `self.ready`?

Reply:
Changed. Although I'm not sure I've done anything special to make sure the self.ready is inherited anywhere. This is just on the encoder side. Not the model side.

Reply:
Resolving conversation to shift it to under Lucien's comment.